kotlinx-coroutines-core/jvm/test/ThreadContextMutableCopiesTest.kt

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright 2016-2018 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.coroutines
+
+import kotlin.coroutines.*
+import kotlin.test.*
+
+class ThreadContextMutableCopiesTest : TestBase() {
+    companion object {
+        val threadLocalData: ThreadLocal<MutableList<String>> = ThreadLocal.withInitial { ArrayList() }
+    }
+
+    class MyMutableElement(
+        private val mutableData: MutableList<String>
```

Comment:
Here is the trickier use-case when the element "mirrors" the mutable data.

It may be too grotesque to be actually useful. For example, if the data within the element is never mutable, but always mirrors some thread-local or any other storage -- we can get rid of `merge` completely and just invoke `copyForChildCoroutine()` on the resulting `coroutineContext + context`. WDYT?